keps/sig-node/2823-pod-addmission-plugin/README.md

```diff
@@ -0,0 +1,445 @@
+a host.
+
+In building the AWS EKS support for AWS Fargate, the AWS Kubernetes engineering
+team faced a dilemma: how could they prevent Pods destined to run on Fargate
+nodes from using host networking or assuming elevated host user privileges? This
+is required to restrict pods which are not supported on these worker nodes and
+also to prevent pods from accessing some portion of file system on the node or
+tampering other process on the host.
+
+The team initially investigated using a webhook or Pod Security Policy (PSP)
+that would prevent Pods with a Fargate scheduler type from having an elevated
+security context or using host networking. However, because the EKS user has
+administrative rights on the Kubernetes cluster, API-layer constructs such as a
+Pod Security Policy or validation webhook may be deleted, which would
+effectively disable the effect of that PSP. Likewise, the second solution the
+team landed on — using Node taints and tolerations — was similarly bound to the
+Kubernetes API layer, which meant EKS users could modify those Node taints and
+tolerations, effectively disabling the effects. A third potential solution
+involving OCI hooks was then investigated. OCI hooks are separate executables
+that an OCI-compatible container runtime invokes that can modify the behaviour
+of the containers in a sandbox. While this solution would have solved the
+API-layer problem, it introduced other issues, such as the inefficiency of
+downloading the container image to the Node before the OCI hook was run. Since
+pods can be scheduled directly on the node using nodename parameter in pod spec,
+its required to add such validation in Kubelet itself.
+
+The final solution the EKS team settled on involved changing kubelet itself to
+support additional node-local Pod admission checks. This KEP outlines the EKS
+team’s approach and proposes upstreaming these changes to kubelet in order to
+allow extensible node-local last-minute validation checks. This functionality
+will enable cloud providers who support nodeless/serverless worker nodes to
+restrict Pods based on fields other than those already being validated by
+kubelet.
+
+Also, Kubernetes scheduler is extensible but isn’t omniscient about everything
+on the node hence there can be cases where a scheduling decision cannot be
+fulfilled by the node based on the current node conditions such as node-level
+resource allocation. The in-tree PodAdmitHandlers mentioned below are good
+examples for this.
+
+* Eviction manager rejects pods if the node is under high pressure (e.g., PIDs)
+* Topology manager rejects pods if the NUMA affinity cannot be satisfied
+
+There can be additional resource allocation conditions and policies handled
+externally to the kubelet, e.g.,
+
+* Alternative resource allocator implemented by the CRI which needs to reject
+  pods in conditions similar to the built-in resource managers.
+
+These can also benefit from a pluggable admission handler to reject pods before
```

Comment:
I think a key question this KEP doesn't really clarify is whether this set of hooks applies ONLY to things that are representable in the API model (things that show up in `Pod`) or whether it applies to other parts of the Kubelet logic that are internal to how Kubelet models pods (the runtime pod, volumes, csi driver interactions).  If it's the former, you're effectively duplicating apiserver admission (literally pod admission hooks take a `Pod` and say yes or no), and you should probably simply leverage that mechanism - "A user should be able to reject a Pod for admission at the Kubelet and the mechanism should look identical to rejecting a pod for admission at the APIserver".  If it's the latter, we would need to define formal APIs for those, and if we do that those APIs should probably still look like an apiserver admission (if we have an API, it should be kubelike, if it's kubelike, the existing webhook pattern solves it, so let's not invent another pattern).

A second concern as well is whether mutation is allowed, or whether this is only validation.  If it's mutation, some of the implications of the apiserver apply and should be addressed (and it's also an even stronger argument that if you want kubelet admission hooks, it should look like apiserver hooks).

I think before we get into the guts of discussing "how you register these" we need to address the questions above.